src/services/DomainModel.ts

```diff
@@ -0,0 +1,48 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a Query is probably in effect
+// which has limited the extent to which we have populated all the objects into
+// memory. For example, if you're working with recent data from the  'Black
+// Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export interface Unknown {
+  // Unknown means you've reached an unexplored part of the tree, not a dead end.
```

Comment:
Great that we aren't mixing in lazy loading, which should be mostly transparent. 👍 

I'm still puzzled why we need Unknown.  The discussion doesn't have to be about a database but any source of data - notehub API, graph QL, etc. and we should specifically take into account the needs of the consumer of the data, i.e. the app frontend and a REST API.   How is Unknown used in those contexts?  If you could work through an example that would help. 

Query interfaces by their nature only return a portion of the model. E.g. a query to notehub to retrieve events in a specified date range only returns those events without a "Unknown" sentinel to mark the end of the data and only for the devices requested. (Pagination is a different story, and the result set can include metadata that describes how to fetch the next/previous page of data based on the original query, and whether those pages exist.)

A worked through example may help change my perspective, but presently I feel we should focus on what we know we need. If we need to build out the model to accommodate the unexplored edges of the dataset we can do that later imho.

Reply:
Simple example: in order to fetch a Gateway without fetching Nodes we need something like this

```ts
export interface Gateway {
  name: string;
  nodes: Set<Node> | Unknown;
}
```

or

```ts
export interface Gateway {
  name: string;
  nodes?: Set<Node>; // ? means `| undefined`
}
```

Which is almost as good. What do you think? Maybe it's better. I just liked inventing my own type for my own semantics instead of reusing `undefined` which could also have the semantics of `nobody pared any nodes with this gateway yet` or myriad similar semantics. Or am I being paranoid and every new typescript programmer will understand that `undefined` means `meh, we don't currently know, but there might be an answer if we looked hard enough`?

Reply:
Javascript has `undefined`, which is indeterminate.   `null` means none for sure.   I see we are leaning towards a flexible interface where the result type moulds to the type of query according to the query input.  `nodes?` I think is pretty standard. 

If the query doesn't ask for it, it doesn't need to be in the result, and then `undefined` is perfectly fine and also well known. If it's asked for and there are now results, an empty collection or `null` is (somewhat) fine. Empty collection is preferred to reinforce the emptiness of the result set.

I don't think there's need for anything else. We are inventing new things, yet doing something quite routine.  :-)